src/dataflow/src/render/reduce.rs

```diff
@@ -187,163 +194,113 @@ where
-                                // The inputs are triples of a reduction type, an optional index and row to decode.
-                                // The `reduce_abelian` operator guarantees that values will be placed in sorted order
-                                // by the values Ord implementation. This means that fused reductions
-                                // (ReductionType's FusedAccumulable and FusedHierarchical) will be the first entries in inputs
-                                // if any such ReductionType's are present.
+                            move |key, input, output| {
+                                // The inputs are pairs of a reduction type, and a row consisting of densely packed fused
+                                // aggregate values.
                                 // We need to reconstitute the final value by:
-                                // 1. Extracting out the fused aggregates
-                                // 2. For each aggregate we are computing, figure out if it is either a fused accumulable,
-                                //    fused hierarchical or unfused aggregate
-                                // 3. Get the relevant value (either from the rows of fused aggregates, or from the input
-                                //    array for unfused aggregates)
-                                // 4. Stitch all the values together into one row.
+                                // 1. Extracting out the fused rows by type
+                                // 2. For each aggregate, figure out what type it is, and grab the relevant value
+                                //    from the corresponding fused row.
+                                // 3. Stitch all the values together into one row.
 
+                                // TODO change these to be options probably
                                 let mut accumulable = DatumList::empty().iter();
                                 let mut hierarchical = DatumList::empty().iter();
-                                let mut input_to_skip = None;
-
-                                // Extract out the fused accumulable and hierarchical aggregates from our inputs if we have them. These
-                                // should all be sorted before any unfused aggregates, so that we can grab them efficiently and at the
-                                // extract the subsets of inputs that corresponds to only unfused aggregates.
-                                for i in 0..input.len() {
-                                    match (input[i].0).0 {
-                                        ReductionType::FusedAccumulable => {
-                                            accumulable = (input[i].0).2.iter();
-                                            input_to_skip = Some(i);
+                                let mut basic = DatumList::empty().iter();
+
+                                // TODO assert that there are <= 3 items?
```

Comment:
would appreciate second opinions on this

Reply:
Ok so I decided to add a log line when we get negative multiplicities in the input vector -- I'm not 100% sure if thats right @frankmcsherry wdyt?